src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/MongoStockRecordRepository.kt

```diff
@@ -27,73 +32,84 @@ import org.springframework.data.mongodb.core.mapping.Document as SpringDocument
 
 @Repository
 class MongoStockRecordRepository(
-    private val mongoTemplate: MongoTemplate,
+    private val reactiveMongoTemplate: ReactiveMongoTemplate,
 ) : StockRecordRepository {
-    override fun insertAll(mongoStockRecords: List<MongoStockRecord>): List<MongoStockRecord> {
-        return mongoTemplate.insertAll(mongoStockRecords).toList()
+    override fun insertAll(mongoStockRecords: List<MongoStockRecord>): Flux<MongoStockRecord> {
+        return reactiveMongoTemplate.insertAll(mongoStockRecords)
     }
 
-    override fun findAllStockSymbols(): List<String> {
-        return mongoTemplate.getCollection(MongoStockRecord.COLLECTION_NAME)
-            .distinct(MongoStockRecord::symbol.name, String::class.java)
-            .toList()
+    override fun findAllStockSymbols(): Flux<String> {
+        return reactiveMongoTemplate.query(MongoStockRecord::class.java)
+            .distinct(MongoStockRecord::symbol.name)
+            .asType<String>()
+            .all()
     }
 
     override fun findTopNStockSymbolsWithStockRecords(
         n: Int,
         from: Date,
         to: Date,
-    ): Map<String, List<MongoStockRecord>> {
-        val maxChange = getMaxBigDecimal(MongoStockRecord::change.name, from, to)
-        val maxPercentChange = getMaxBigDecimal(MongoStockRecord::percentChange.name, from, to)
-
-        val matchOperation: MatchOperation =
-            Aggregation.match(
-                Criteria.where(MongoStockRecord::dateOfRetrieval.name)
-                    .gte(from)
-                    .lt(to)
-            )
-        val groupOperation: GroupOperation = Aggregation.group(MongoStockRecord::symbol.name)
-            .push("$\$ROOT").`as`("records")
-        val projectOperation = getProjectWithAvgMaxValues(maxChange, maxPercentChange)
-        val weightedPipeline = getWeightedPipeLine()
-
-        val aggregation = Aggregation.newAggregation(
-            matchOperation,
-            groupOperation,
-            projectOperation,
-            weightedPipeline,
-            Aggregation.sort(Sort.by("weight").descending()),
-            Aggregation.limit(n.toLong())
-        )
-
-        val results: List<ResultingClass> = mongoTemplate.aggregate(
-            aggregation,
-            MongoStockRecord.COLLECTION_NAME,
-            ResultingClass::class.java
-        ).mappedResults
+    ): Mono<Map<String, List<MongoStockRecord>>> {
+        val maxChangeMono = getMaxBigDecimal(MongoStockRecord::change.name, from, to)
+        val maxPercentChangeMono = getMaxBigDecimal(MongoStockRecord::percentChange.name, from, to)
+
+        return Mono.zip(maxChangeMono, maxPercentChangeMono)
+            .flatMap { maxValues ->
```

Comment:
Try to keep your reactor chain small inside. Here you have a lot of code inside flatMap and it is unclear what operation is blocking. Extract this code to separate function, please